datafusion/expr/src/logical_plan/builder.rs

```diff
@@ -545,14 +545,22 @@ impl LogicalPlanBuilder {
 
                     match (&l.relation, &r.relation) {
                         (Some(lr), Some(rr)) => {
-                            let l_is_left =
-                                self.plan.schema().field_with_qualified_name(lr, &l.name);
-                            let l_is_right =
-                                right.schema().field_with_qualified_name(lr, &l.name);
-                            let r_is_left =
-                                self.plan.schema().field_with_qualified_name(rr, &r.name);
-                            let r_is_right =
-                                right.schema().field_with_qualified_name(rr, &r.name);
+                            let l_is_left = self.plan.schema().field_with_qualified_name(
+                                &lr.as_table_reference(),
```

Comment:
I tried playing around with the signature of `field_with_qualified_name` and I couldn't find a way to make `lr` still work. 🤷

Reply:
Yeah was thinking maybe having something like `Into<TableReference>` or the like.

Related, but I feel TableReference and its related structs OwnedTableReference and ResolvedTableReference got a bit out of hand, since they seem all very similar and have conversions between them, but still are quite separate. I think I initially had `field_with_qualified_name` accept a reference to an OwnedTableReference but changed it to TableReference as that wouldn't require owning a bunch of strings if you just wanted to pass in a reference, but can see it doesn't really result in a nice API imo

Maybe need refactoring around the TableReference structs to try make them more interchangeable, via a common trait possibly

Reply:
> Maybe need refactoring around the TableReference structs to try make them more interchangeable, via a common trait possibly

I wonder if we could maybe change TableReference to use `Cow` somehow rather than separate variants 🤔

Reply:
How do you mean?

Reply:
I was thinking about removing OwnedTableReference -- I started to sketch out what this might look like in https://github.com/apache/arrow-datafusion/pull/5508

Working with the structures then I think becomes less cumbersome